File: src/parser/frontmatter.ts
import { err, ok, type Result } from "../lib/result.ts";
import {
  FrontmatterRelationshipsSchema,
  type Relationship,
  type RelationshipType,
} from "../schema/index.ts";

export type FrontmatterData = {
  slug: string;
  title?: string;
  summary?: string;
  aliases?: string[];
  tags: string[];
  relationships: Relationship[];
  body: string;
};

export type FrontmatterError = { type: "frontmatter"; message: string };

const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---\r?\n([\s\S]*)$/;

export function parseFrontmatter(raw: string): Result<FrontmatterData, FrontmatterError> {
  const match = raw.match(FRONTMATTER_RE);
  if (!match) {
    return err({ type: "frontmatter", message: "No frontmatter found" });
  }

  const [, yamlBlock, body] = match;
  const parsed = parseYaml(yamlBlock!);

  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    return err({ type: "frontmatter", message: "Frontmatter is not a valid object" });
  }

  const obj = parsed as Record<string, unknown>;

  if (typeof obj.slug !== "string" || obj.slug.length === 0) {
    return err({ type: "frontmatter", message: "Missing required field: slug" });
  }

  const tags = Array.isArray(obj.tags)
    ? obj.tags.filter((t): t is string => typeof t === "string")
    : [];

  const relationships = normalizeRelationships(obj.relationships);

  const summary =
    typeof obj.summary === "string" && obj.summary.length > 0 ? obj.summary : undefined;
  const aliases = Array.isArray(obj.aliases)
    ? obj.aliases.filter((a): a is string => typeof a === "string")
    : undefined;

  return ok({
    slug: obj.slug,
    title: typeof obj.title === "string" ? obj.title : undefined,
    summary,
    aliases: aliases && aliases.length > 0 ? aliases : undefined,
    tags,
    relationships,
    body: body!,
  });
}

function normalizeRelationships(raw: unknown): Relationship[] {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return [];

  const parsed = FrontmatterRelationshipsSchema.safeParse(raw);
  if (!parsed.success) return [];

  const result: Relationship[] = [];
  for (const [type, targets] of Object.entries(parsed.data) as [
    RelationshipType,
    string | string[],
  ][]) {
    const targetList = Array.isArray(targets) ? targets : [targets];
    for (const target of targetList) {
      result.push({ target, type });
    }
  }
  return result;
}

export function parseYaml(yaml: string): unknown {
  const result: Record<string, unknown> = {};
  const lines = yaml.split("\n");
  let currentKey: string | null = null;
  let currentArray: string[] | null = null;

  for (const line of lines) {
    if (line.trim() === "" || line.trim().startsWith("#")) continue;

    const keyMatch = line.match(/^(\w[\w-]*):\s*(.*)/);
    if (keyMatch) {
      if (currentKey && currentArray) {
        result[currentKey] = currentArray;
        currentArray = null;
      }

      const [, key, value] = keyMatch;
      currentKey = key!;

      const trimmed = value!.trim();

      if (trimmed === "") {
        currentArray = null;
      } else if (trimmed.startsWith("[") && trimmed.endsWith("]")) {
        const inner = trimmed.slice(1, -1);
        result[currentKey] = inner
          .split(",")
          .map((s) => s.trim())
          .filter((s) => s.length > 0);
        currentKey = null;
      } else {
        result[currentKey] = stripQuotes(trimmed);
        currentKey = null;
      }
      continue;
    }

    const nestedKeyMatch = line.match(/^\s{2}(\w[\w-]*):\s*(.*)/);
    if (nestedKeyMatch && currentKey) {
      if (currentArray) {
        result[currentKey] = currentArray;
        currentArray = null;
      }
      if (typeof result[currentKey] !== "object" || Array.isArray(result[currentKey])) {
        result[currentKey] = {};
      }
      const [, nestedKey, nestedValue] = nestedKeyMatch;
      const trimmed = nestedValue!.trim();

      if (trimmed.startsWith("[") && trimmed.endsWith("]")) {
        (result[currentKey] as Record<string, unknown>)[nestedKey!] = trimmed
          .slice(1, -1)
          .split(",")
          .map((s) => s.trim())
          .filter((s) => s.length > 0);
      } else {
        (result[currentKey] as Record<string, unknown>)[nestedKey!] = trimmed;
      }
      continue;
    }

    const arrayItemMatch = line.match(/^\s*-\s+(.*)/);
    if (arrayItemMatch && currentKey) {
      if (!currentArray) currentArray = [];
      currentArray.push(arrayItemMatch[1]!.trim());
    }
  }

  if (currentKey && currentArray) {
    result[currentKey] = currentArray;
  }

  return result;
}

function stripQuotes(value: string): string {
  if (value.length >= 2) {
    const first = value[0];
    const last = value[value.length - 1];
    if ((first === '"' && last === '"') || (first === "'" && last === "'")) {
      return value.slice(1, -1);
    }
  }
  return value;
}

File: crates/pramana-parser/src/frontmatter.rs
use std::sync::LazyLock;

use regex::Regex;
use serde_yaml::Value;

use crate::error::ParseError;
use crate::types::{FrontmatterData, Relationship, RelationshipType};

static FRONTMATTER_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?s)^---\r?\n(.*?)\r?\n---\r?\n(.*)$").unwrap());

pub fn parse_frontmatter(raw: &str) -> Result<FrontmatterData, ParseError> {
    let caps = FRONTMATTER_RE
        .captures(raw)
        .ok_or_else(|| ParseError::Frontmatter {
            message: "No frontmatter found".into(),
        })?;

    let yaml_block = &caps[1];
    let body = caps[2].to_string();

    let value: Value = serde_yaml::from_str(yaml_block).map_err(|e| ParseError::Frontmatter {
        message: format!("Invalid YAML: {e}"),
    })?;

    let mapping = value.as_mapping().ok_or_else(|| ParseError::Frontmatter {
        message: "Frontmatter is not a valid object".into(),
    })?;

    let slug = mapping
        .get("slug")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .ok_or_else(|| ParseError::Frontmatter {
            message: "Missing required field: slug".into(),
        })?
        .to_string();

    let tags = mapping
        .get("tags")
        .and_then(Value::as_sequence)
        .map(|seq| {
            seq.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();

    let relationships = mapping
        .get("relationships")
        .map(normalize_relationships)
        .unwrap_or_default();

    let title = mapping
        .get("title")
        .and_then(Value::as_str)
        .map(String::from);

    let summary = mapping
        .get("summary")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(String::from);

    let aliases: Option<Vec<String>> = mapping
        .get("aliases")
        .and_then(Value::as_sequence)
        .map(|seq| {
            seq.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .filter(|a: &Vec<String>| !a.is_empty());

    Ok(FrontmatterData {
        slug,
        title,
        summary,
        aliases,
        tags,
        relationships,
        body,
    })
}

fn normalize_relationships(value: &Value) -> Vec<Relationship> {
    let mapping = match value.as_mapping() {
        Some(m) => m,
        None => return vec![],
    };

    let mut result = Vec::new();
    for (key, val) in mapping {
        let type_str = match key.as_str() {
            Some(s) => s,
            None => continue,
        };

        let rel_type = match type_str.parse::<RelationshipType>() {
            Ok(t) => t,
            Err(()) => continue,
        };

        let targets: Vec<String> = match val {
            Value::String(s) => vec![s.clone()],
            Value::Sequence(seq) => seq
                .iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect(),
            _ => continue,
        };

        for target in targets {
            result.push(Relationship {
                target,
                rel_type: rel_type.clone(),
                line: None,
                section: None,
            });
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_slug_tags_and_relationships() {
        let raw = "---\nslug: order\ntags: [entity, commerce, core]\nrelationships:\n  depends-on: [customer, line-item, shipping-info]\n---\n\n# Order\n\nContent here.";

        let result = parse_frontmatter(raw);
        assert!(result.is_ok());
        let data = result.unwrap();

        assert_eq!(data.slug, "order");
        assert_eq!(data.tags, vec!["entity", "commerce", "core"]);
        assert_eq!(data.relationships.len(), 3);
        assert!(data
            .relationships
            .iter()
            .any(|r| r.target == "customer" && r.rel_type == RelationshipType::DependsOn));
        assert!(data
            .relationships
            .iter()
            .any(|r| r.target == "line-item" && r.rel_type == RelationshipType::DependsOn));
        assert!(data
            .relationships
            .iter()
            .any(|r| r.target == "shipping-info" && r.rel_type == RelationshipType::DependsOn));
    }

    #[test]
    fn returns_error_when_no_frontmatter() {
        let result = parse_frontmatter("# Just a heading\n\nNo frontmatter.");
        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(matches!(err, ParseError::Frontmatter { .. }));
    }

    #[test]
    fn returns_error_when_slug_is_missing() {
        let raw = "---\ntags: [test]\n---\n\nContent.";
        let result = parse_frontmatter(raw);
        assert!(result.is_err());
        let err = result.unwrap_err();
        match err {
            ParseError::Frontmatter { message } => assert!(message.contains("slug")),
            _ => panic!("expected frontmatter error"),
        }
    }

    #[test]
    fn handles_empty_tags() {
        let raw = "---\nslug: minimal\n---\n\n# Minimal";
        let result = parse_frontmatter(raw).unwrap();
        assert_eq!(result.tags, Vec::<String>::new());
        assert_eq!(result.relationships, Vec::<Relationship>::new());
    }

    #[test]
    fn extracts_title_from_frontmatter() {
        let raw = "---\nslug: test\ntitle: Custom Title\n---\n\n# Different Title";
        let result = parse_frontmatter(raw).unwrap();
        assert_eq!(result.title, Some("Custom Title".into()));
    }

    #[test]
    fn preserves_body_content() {
        let raw = "---\nslug: test\n---\n\n# Title\n\nBody content here.";
        let result = parse_frontmatter(raw).unwrap();
        assert!(result.body.contains("Body content here."));
    }

    #[test]
    fn extracts_summary_from_frontmatter() {
        let raw =
            "---\nslug: order\nsummary: \"A customer's intent to purchase one or more products\"\n---\n\n# Order";
        let result = parse_frontmatter(raw).unwrap();
        assert_eq!(
            result.summary,
            Some("A customer's intent to purchase one or more products".into())
        );
    }

    #[test]
    fn extracts_aliases_from_frontmatter() {
        let raw =
            "---\nslug: order\naliases: [purchase-order, sales-order, transaction]\n---\n\n# Order";
        let result = parse_frontmatter(raw).unwrap();
        assert_eq!(
            result.aliases,
            Some(vec![
                "purchase-order".into(),
                "sales-order".into(),
                "transaction".into()
            ])
        );
    }

    #[test]
    fn summary_and_aliases_are_none_when_absent() {
        let raw = "---\nslug: minimal\n---\n\n# Minimal";
        let result = parse_frontmatter(raw).unwrap();
        assert_eq!(result.summary, None);
        assert_eq!(result.aliases, None);
    }

    #[test]
    fn strips_quotes_from_summary_value() {
        let raw = "---\nslug: test\nsummary: 'Single quoted summary'\n---\n\nBody.";
        let result = parse_frontmatter(raw).unwrap();
        assert_eq!(result.summary, Some("Single quoted summary".into()));
    }
}
